### ISSUE TITLE
Support for `MetaProcessor::process`with values overloading the `&` operator.

### ISSUE THREAD

**Issue by SylvainCorlay** [vgvassilev/cling#176]
Calling `MetaProcessor::process` with a non-`nullptr` value causes compilation error in the interpreted code, namely `No matching function for call to setValueNoAlloc` in the case where the returned value overloads the `&` operator.

```cpp
class notapointer {};

struct Bar
{
    inline notapointer operator&()
    {
        return notapointer();
    }
};
```

**Comment by SylvainCorlay** [comment]
I suspect the culprit is

```cpp
        // we need to get the address of the references
        Expr* AddrOfE = m_Sema->BuildUnaryOp(/*Scope*/0, noLoc, UO_AddrOf,
                                             E).get();
        CallArgs.push_back(AddrOfE);
```

`UO_AddrOf` probably makes use of `operator&`. If this is it, a better solution would be to use `std::addressof`.

**Comment by vgvassilev** [comment]
What's the call in cling that breaks it? Is it just `Bar b`?

**Comment by SylvainCorlay** [comment]
Yes, but you would need to change the UserInterface's call to process and pass it a value.

```
Value v;
const int indent = m_MetaProcessor->process(Line, compRes, &v);
```

**Comment by SylvainCorlay** [comment]
If there was another way to get the expression for the address of an object than `Expr* AddrOfE = m_Sema->BuildUnaryOp(/*Scope*/0, noLoc, UO_AddrOf, E).get();`, this is probably what should be used.

**Comment by SylvainCorlay** [comment]
I can work around the bug with the following patch to `ValueExtractionSynthesizer` in the mean time.

```diff
--- a/lib/Interpreter/ValueExtractionSynthesizer.cpp
+++ b/lib/Interpreter/ValueExtractionSynthesizer.cpp
@@ -348,7 +348,14 @@ namespace {
         // we need to get the address of the references
         Expr* AddrOfE = m_Sema->BuildUnaryOp(/*Scope*/0, noLoc, UO_AddrOf,
                                              E).get();
-        CallArgs.push_back(AddrOfE);
+
+        QualType AddrOfETy = (AddrOfE) ? AddrOfE->getType() : m_Context->VoidTy;
+        QualType desugaredTyA = AddrOfETy.getDesugaredType(*m_Context);
+
+        if (desugaredTyA->isAnyPointerType())
+        {
+            CallArgs.push_back(AddrOfE);
+        }
       }
       else if (desugaredTy->isAnyPointerType()) {
         // function pointers need explicit void* cast.
@@ -375,11 +382,6 @@ namespace {
         Call = m_Sema->ActOnCallExpr(/*Scope*/0, m_UnresolvedNoAlloc,
                                    locStart, CallArgs, locEnd);
       }
-      else {
-        m_Sema->Diag(locStart, diag::err_unsupported_unknown_any_decl) <<
-          utils::TypeName::GetFullyQualifiedName(desugaredTy, *m_Context) <<
-          SourceRange(locStart, locEnd);
-      }
     }
```

**Comment by vgvassilev** [comment]
This should trigger the problem in cling standalone by:
```
class notapointer {};

struct Bar {
    inline notapointer operator&() {
        return notapointer();
    }
};

Bar b
```

**Comment by SylvainCorlay** [comment]
Cling standalone's user interface does not call `process` with a non-`nullptr` value as third argument, so there is no error.

**Comment by vgvassilev** [comment]
Go ahead and open a PR. Why did you need to remove the `else` part?

**Comment by vgvassilev** [comment]
Go ahead and paste that minimal example and you will see the same problem you reported. Internally, cling uses `cling::Value` when you omit the final `;`.

**Comment by SylvainCorlay** [comment]
> Go ahead and paste that minimal example and you will see the same problem you reported. Internally, cling uses cling::Value when you omit the final ;.

Ah, yes, I was using my modified build 😄 

> Why did you need to remove the else part?

Because I am just completely dropping the argument in the case of such an object *for now*, and it makes it fall into the unhandled type case.

**Comment by SylvainCorlay** [comment]
All we need for a real fix is a means to get a replacement for

```
Expr* AddrOfE = m_Sema->BuildUnaryOp(/*Scope*/0, noLoc, UO_AddrOf, E).get();
```

I opened a stackoverflow issue here:
https://stackoverflow.com/questions/49032835/clangs-uo-addrof-unary-operator-sensible-to-overloading-of-operator

**Comment by vgvassilev** [comment]
Let me take care of it. I have minimal cling-standalone reproducer and this is all I need.

**Comment by SylvainCorlay** [comment]
Thanks, this is going to be super useful!
We use a lot of `operator&` overloading for pointer proxies.

**Comment by SylvainCorlay** [comment]
Hum, I have tried diving into clang's internals to fix this, but with no luck so far.

**Comment by vgvassilev** [comment]
@SylvainCorlay thanks for the report. It should be fixed in 01dbb6d

**Comment by SylvainCorlay** [comment]
Awesome, thanks!
